compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1446,7 +1446,9 @@ trait Applications extends Compatibility {
             unapplyArgType
 
         val dummyArg = dummyTreeOfType(ownType)
-        val unapplyApp = typedExpr(untpd.TypedSplice(Apply(unapplyFn, dummyArg :: Nil)))
+        val unapplyApp = withMode(Mode.NoInline) { // transparent inline unapplies are exapnded when indexing the pattern see `indexPattern` in Typer.
```

Comment:
This will not work. We need to refine the types for the transparent unapply at this point. 

We might need to fully inline the call here, which is proving to be a bit more involved.

Reply:
I managed to make it work.

The trick is to type this application with the dummy argument without inlining a just after that inline the unapply function. We also need to adapt unapply function `unapplyFn` to make the new code use the inlined code. Note that the inlining of the unapply function must be done on the function that receives the dummy argument, trailing implicits are kept as is.